### PR TITLE
webgpu: add a naive implementation of conv3d

### PR DESCRIPTION
**Pull request by xhcao**
To see the logs from the Cloud Build CI, please join either our [discussion](https://groups.google.com/a/tensorflow.org/forum/#!forum/tfjs) or [announcement](https://groups.google.com/a/tensorflow.org/forum/#!forum/tfjs-announce) mailing list.

- - -
This change is [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs/7016)

### CODE REVIEW

**Comment by xhcao** [comment]
Hi, @qjia7 @gyagp Some other kernels also depend on conv3d kernel, so I firstly added a naive conv3d implementation here, in order to quickly implement the other kernels. I will to implement a shared conv3d version in the future.

**Review by gyagp — APPROVED** [review]
LGTM with some nits. Overall, we should use more WGSL syntax sugar, like ++ and +=.

File: tfjs-backend-webgpu/src/shader_util.ts [hunk 1]
```diff
@@ -18,12 +18,13 @@
...
     indicesArr: number[], variableName: string): string[] {
-  if (Math.max(...indicesArr) > 3) {
-    throw new Error('Cannot symbolically compute strides for rank > 4 tensor.');
+  if (Math.max(...indicesArr) > 5) {
+    throw new Error('Cannot symbolically compute strides for rank > 6 tensor.');
   }
 
   const numCoords = indicesArr.length;
-  const shape = indicesArr.map(d => `${variableName}[${d}]`);
+  const indicesStr = ['.x', '.y', '.z', '.w', '.u', '.v'];
```
[the remark this answers is not in the document]
Comment (xhcao) [reply tfjs-backend-webgpu/src/shader_util.ts]: Hi, @gyagp, xyzw is used for single component selection by WGSL https://gpuweb.github.io/gpuweb/wgsl/#:~:text=7.7.1.1.%20Vector%20Single%20Component%20Selection
TFJS webgpu backend extends the vec4 to vec5 and vec6 https://github.com/tensorflow/tfjs/blob/ea8d7db6d050dbdde3b1586b88c6d620890dadb4/tfjs-backend-webgpu/src/webgpu_program.ts#L313, and uses `uv` to select the fifth and sixth components.

All other comments are addressed, thank you.

File: tfjs-backend-webgpu/src/conv3d_naive_webgpu.ts [hunk 2]
```diff
@@ -0,0 +1,129 @@
+/**
+ * @license
+ * Copyright 2022 Google LLC.
```
Comment (qjia7) [inline tfjs-backend-webgpu/src/conv3d_naive_webgpu.ts:3]: nit: `2022` -> `2023` and similar for other places

File: tfjs-backend-webgpu/src/conv3d_naive_webgpu.ts [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+
+export class Conv3DNaiveProgram implements WebGPUProgram {
+  outputShape: number[];
+  shaderKey: string;
+  dispatchLayout: {x: number[], y: number[], z: number[]};
+  dispatch: [number, number, number];
+  variableNames = ['x', 'W'];
+  uniforms =
+      'filterDims: vec3<i32>, pad: vec3<i32>, strides: vec3<i32>, dilations: vec3<i32>,';
+  workgroupSize: [number, number, number] = [4, 4, 8];
```
Comment (qjia7) [inline tfjs-backend-webgpu/src/conv3d_naive_webgpu.ts:31]: Can we use the flat dispatch layout here? Previous experience shows that flat dispatch layout always has good performance if the algorithm is irrelevant with dispatch layout. You can have a micro-benchmark to verify it.

**Review by qjia7 — APPROVED** [review]

File: tfjs-backend-webgpu/src/conv3d_naive_webgpu.ts (hunk 3 above)
Comment (xhcao) [reply tfjs-backend-webgpu/src/conv3d_naive_webgpu.ts]: Done

File: tfjs-backend-webgpu/src/shader_util.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (gyagp) [reply tfjs-backend-webgpu/src/shader_util.ts]: Thanks for the info!

**Pull request merged**